fuzzers/main.py

```diff
@@ -0,0 +1,450 @@
+import subprocess
+import sys
+import threading
+import time
+import toml
+import typing
+import uuid
+
+from lib import config
+from workers import utils
+
+WORKDIR = utils.WORKDIR
+REPO_URL = utils.REPO_URL
+#WORKDIR = pathlib.Path('/tmp')
+#REPO_URL = 'https://github.com/Ekleog/nearcore'
+def connect_to_gcs() -> google.cloud.storage.Client:
+    """Setup the environment to have gsutils work, and return a connection to GCS"""
+    subprocess.run(
+        ['gcloud', 'auth', 'activate-service-account', '--key-file', GCS_CREDENTIALS_FILE],
+        check = True,
+    )
+    return google.cloud.storage.Client.from_service_account_json(str(GCS_CREDENTIALS_FILE))
+    #return google.cloud.storage.Client(project = 'near-nayduck')
+
+NUM_FUZZERS = typing.cast(int, os.cpu_count())
+#NUM_FUZZERS = 1
+
+AUTO_REFRESH_INTERVAL_SECS = 24 * 3600
+#AUTO_REFRESH_INTERVAL_SECS = 300
+
+CMD_PORT = 7055
+METRICS_PORT = 5507
+
+REPO_DIR_PARENT = WORKDIR
+REPO_DIR_NAME = 'fuzzed-nearcore'
+REPO_DIR = REPO_DIR_PARENT / REPO_DIR_NAME
+
+GCS_CREDENTIALS_FILE = config.CONFIG_DIR / 'credentials.json'
+GCS_BUCKET = 'fuzzer'
+
+# pylint: disable=line-too-long
+FUZZ_TIME = prometheus_client.Counter('fuzz_seconds', 'Time spent fuzzing', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CRASHES = prometheus_client.Counter('fuzz_crashes', 'Number of times the fuzzer process crashed (not unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_ARTIFACTS_FOUND = prometheus_client.Counter('fuzz_artifacts_found', 'Number of artifacts found (should be number of unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_UPLOADED = prometheus_client.Counter('fuzz_corpus_uploaded', 'Number of elements uploaded to GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_DELETED = prometheus_client.Counter('fuzz_corpus_deleted', 'Number of elements deleted from GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+# pylint: enable=line-too-long
+
+class FuzzProcess:
+    def __init__(
+        self,
+        crate: str,
+        runner: str,
+        proc: subprocess.Popen,
+        exit_event: threading.Event,
+    ):
+        """Create a FuzzProcess object"""
+        self.crate = crate
+        self.runner = runner
+        self.proc = proc
+        self.exit_event = exit_event
+
+def update_repo(branch: str) -> None:
+    """Update the repository at REPO_DIR to the tip of github's branch `branch`"""
+    if not REPO_DIR.exists():
+        print(f"Doing initial clone of repository {REPO_DIR}")
+        subprocess.check_call(['git', 'clone', REPO_URL, REPO_DIR_NAME], cwd=REPO_DIR_PARENT)
+
+    print(f"Updating to latest commit of branch {branch}")
+    subprocess.check_call(['git', 'fetch', REPO_URL, branch], cwd=REPO_DIR)
+    subprocess.check_call(['git', 'checkout', 'FETCH_HEAD'], cwd=REPO_DIR)
+
+# TODO: make these types more precise? But then everything comes from toml.load so...
+ConfigType = typing.Any
+BranchType = typing.Any
+TargetType = typing.Any
```

Comment:
I usually go with `typing.Dict[str, typing.Any]` which is slightly more specific.